### PR TITLE
Add Dockerfile

### PR DESCRIPTION
**Pull request by c2h5oh**
Adds a Dockerfile building a light image containing LedisDB with Snappy, RocksDB and LevelDB.

Depends on #326 and should be merged after it. 

Assumes version 0.6 has been tagged during build process

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ENV LEVELDB_VERSION 47cb9e2a211e1d7157078ba7bab536beb29e56dc
+ENV ROCKSDB_VERSION 5.8.6
+ENV GOSU_VERSION 1.10
+
+
+
+WORKDIR /build
+
+RUN apt-get update && \
+	apt-get install -y \
```
Comment (siddontang) [inline Dockerfile]: need alignment here

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+RUN wget -O ledisdb-src.tar.gz "https://github.com/siddontang/ledisdb/archive/$LEDISDB_VERSION.tar.gz" && \
+    tar -zxf ledisdb-src.tar.gz && \
+    mkdir -p /go/src/github.com/siddontang/ && \
+    mv ledisdb-$LEDISDB_VERSION /go/src/github.com/siddontang/ledisdb
+
+# build LevelDB
+RUN wget -O leveldb-src.tar.gz "https://github.com/google/leveldb/archive/$LEVELDB_VERSION.tar.gz" && \
+    tar -zxf leveldb-src.tar.gz && \
+    cd leveldb-$LEVELDB_VERSION && \
+    patch -p0 < /go/src/github.com/siddontang/ledisdb/tools/leveldb.patch && \
```
Comment (siddontang) [inline Dockerfile]: do we need this patch now? I think we can remove it if it can't improve too much performance.

Comment (c2h5oh) [reply Dockerfile]: In your fork you've made a total of 4 changes to LevelDB:
1. you've increased the target (max) file size - which is now exposed as a setting and I've made it configurable in the other branch
2. you've doubled the soft limit on 0-level files (8 -> 16)
3. you've increased the hard cap on 0-level files from 12 to 64
4. you've doubled the grandparent overlap (10 -> 20 times target file size)

Change 1 is no longer required as that value is now exposed in LevelDB and I've made it configurable in Ledis. Changes 2-4 are made using the patch file.

Note: I know next to nothing about LevelDB - all I did was replicating your changes in v0.5.

Comment (siddontang) [reply Dockerfile]: Got it. 

I think we can still keep the patch :-)

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+ENV CGO_LDFLAGS "-L/build/lib -lsnappy"
+
+#build LedisDB
+RUN export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/build/lib:/usr/lib && \
+    export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:/build/lib:/usr/lib && \
+    mkdir -p /build/bin && \
+    rm -rf /build/bin/* && \
+    cd /go/src/github.com/siddontang/ledisdb && \
+    GOGC=off go build -i -o /build/bin/ledis-server -tags "snappy leveldb rocksdb" cmd/ledis-server/* && \
+	GOGC=off go build -i -o /build/bin/ledis-cli -tags "snappy leveldb rocksdb" cmd/ledis-cli/* && \
```
Comment (siddontang) [inline Dockerfile]: make alignment

File: entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+# first arg is `-f` or `--some-option`
+# or first arg is `something.toml`
+if [ "${1#-}" != "$1" ] || [ "${1%.toml}" != "$1" ]; then
+	set -- /bin/ledis-server "$@"
+fi
+
+# allow the container to be started with `--user`
+if [ "$1" = 'ledis-server' -a "$(id -u)" = '0' ]; then
+	chown -R ledis /datastore
```
Comment (siddontang) [inline entrypoint.sh]: make alignment

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+RUN groupadd -r ledis && \
+    useradd -r -g ledis ledis && \
+    mkdir /datastore && \
+    chown ledis:ledis /datastore && \
+    chmod 444 /config.toml && \
+    gosu nobody true
+
+RUN apt-get update && \
+	apt-get install -y --no-install-recommends libsnappy1v5 && \
```
Comment (siddontang) [inline Dockerfile]: make alignment

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	GOGC=off go build -i -o /build/bin/ledis-cli -tags "snappy leveldb rocksdb" cmd/ledis-cli/* && \
+	GOGC=off go build -i -o /build/bin/ledis-benchmark -tags "snappy leveldb rocksdb" cmd/ledis-benchmark/* && \
+	GOGC=off go build -i -o /build/bin/ledis-dump -tags "snappy leveldb rocksdb" cmd/ledis-dump/* && \
+	GOGC=off go build -i -o /build/bin/ledis-load -tags "snappy leveldb rocksdb" cmd/ledis-load/* && \
+	GOGC=off go build -i -o /build/bin/ledis-repair -tags "snappy leveldb rocksdb" cmd/ledis-repair/*
+
+# grab gosu for easy step-down from root
+# https://github.com/tianon/gosu/releases
+RUN set -ex; \
+	dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"; \
```
Comment (siddontang) [inline Dockerfile]: make alignment

**Comment by c2h5oh** [comment]
Indentation fixed.

**Comment by siddontang** [comment]
Thanks @c2h5oh

**Pull request merged**